include/seqan3/alignment/matrix/detail/coordinate_matrix.hpp

```diff
@@ -154,77 +291,128 @@ class coordinate_matrix_iterator
     //!\}
 };
 
-/*!\brief A matrix over coordinates.
- * \ingroup alignment_matrix
- * \implements std::ranges::forward_range
+//------------------------------------------------------------------------------
+// coordinate_matrix_simd_iterator
+//------------------------------------------------------------------------------
+
+/*!\brief The iterator for the vectorised seqan3::detail::coordinate_matrix.
+ * \implements std::forward_iterator
  *
- * \details
+ * \tparam index_t The type of the index; must model seqan3::simd::simd_concept.
  *
- * This matrix emulates a two-dimensional index, such that each cell inside of the alignment matrix can be located
- * through a unique coordinate (=index). In the alignment algorithm this matrix is paired with the alignment matrix
- * to form an indexed alignment matrix.
+ * \details
  *
- * This matrix is cheap as it stores only the dimensions of the matrix and does not allocate any memory for the
- * coordinates. It uses the seqan3::detail::matrix_coordinate_iterator to iterate
- * over the columns of this virtual matrix. When the seqan3::detail::matrix_coordinate_iterator is dereferenced it
- * returns an on-the-fly constructed range ([prvalue](https://en.cppreference.com/w/cpp/language/value_category))
- * representing the current indexed column. The reference type of this range is seqan3::detail::matrix_coordinate.
+ * Iterates over the columns of the underlying seqan3::detail::coordinate_matrix. The iterator returns a transformed
+ * seqan3::views::iota_simd view over the row indices. In the transformation every row index is converted to a
+ * seqan3::detail::matrix_index using the seqan3::detail::convert_to_matrix_coordinate function object.
  */
-class coordinate_matrix
+template <typename index_t>
+//!\cond
+    requires std::integral<index_t> ||
+             (simd::simd_concept<index_t> && requires (simd::simd_traits<index_t> traits)
+             {
+                 requires std::integral<typename decltype(traits)::scalar_type>;
+             })
+//!\endcond
+template <simd_concept simd_index_t>
+class coordinate_matrix<index_t>::coordinate_matrix_simd_iterator
 {
 private:
-    //!\brief Represents the size of a row (number of columns).
-    size_t number_of_columns{};
-    //!\brief Represents the size of a column (number of rows).
-    size_t number_of_rows{};
+    //!\brief The scalar type.
+    using scalar_type = typename simd_traits<simd_index_t>::scalar_type;
+
+    //!\brief The column index as simd vector.
+    simd_index_t simd_column_index{};
+    //!\brief The currently represented column index.
+    scalar_type scalar_column_index{};
```

Comment:
```suggestion
```

I think we already established in another PR that we don't need this extra.

Reply:
jupp, and then we can probably merge this into one implementation with a little if constexpr overhead.